bridge/checkpoint/checkpoint.go

```diff
@@ -0,0 +1,138 @@
+package checkpoint
+
+import (
+	"github.com/0xPolygon/polygon-edge/bridge/checkpoint/transport"
+	"github.com/0xPolygon/polygon-edge/bridge/sam"
+	"github.com/0xPolygon/polygon-edge/network"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+)
+
+const (
+	EpochSize = 100
+)
+
+type Checkpoint interface {
+	Start() error
+	Close() error
+	CreateNewCheckpoint() error
+	SetValidators([]types.Address)
+}
+
+type Blockchain interface {
+	GetBlocks(start, end uint64) ([]types.Block, error)
+}
+
+type checkpoint struct {
+	logger            hclog.Logger
+	signer            sam.Signer
+	blockchain        Blockchain
+	rootchainContract RootChainContractClient
+	transport         transport.CheckpointTransport
+	validators        []types.Address
+
+	closeCh chan struct{}
+}
+
+func NewCheckpoint(
+	logger hclog.Logger,
+	network *network.Server,
+	signer sam.Signer,
+) (Checkpoint, error) {
+	checkpointLogger := logger.Named("checkpoint")
+
+	return &checkpoint{
+		logger: checkpointLogger,
+		signer: signer,
+		// blockchain: blockchain,
+		// rootchainContract: rootchainContractClient,
+		transport: transport.NewLibp2pGossipTransport(logger, network),
+	}, nil
+}
+
+func (c *checkpoint) Start() error {
+	return nil
+}
+
+func (c *checkpoint) Close() error {
+	return nil
+}
+
+func (c *checkpoint) SetValidators(vals []types.Address) {
+	c.validators = vals
+}
+
+func (c *checkpoint) CreateNewCheckpoint() error {
+	// Step1: Get Latest information stored in RootChain contract
+	currentEpoch, err := c.rootchainContract.GetCurrentHeaderBlock()
+	if err != nil {
+		return err
+	}
+
+	lastChildBlock, err := c.rootchainContract.GetLastChildBlock()
+	if err != nil {
+		return err
+	}
+
+	// Step2: Determine the range of next checkpoint and get blocks from local chain
+	start, end := c.determineCheckpointRange(lastChildBlock)
+	blocks, err := c.blockchain.GetBlocks(start, end)
```

Comment:
Is there a chance that some epoch is not going to get inserted in the rootchain contract?
If yes, and let's say our chain is on epoch 7, and the rootchain has been last updated on epoch 3. 
Will our range be from epoch 3 to latest block, or only one epoch (3-4)?

Reply:
Yes that's why we'll have Ack and NoAck mechanism. Besically end will be the latest block number

Reply:
So checkpoint range should be [fromLastChildBlock, toLastBlockInPolygonEdge].

Reply:
Ideally